### PR TITLE
Generate controller stats logs

### PR DESCRIPTION
**Pull request by yiqigao217**
Create counters for controller stats, e.g. number of reconciliation
attemps, or api calls. Generate logs with all the stats if the
enable-test-log flag is set.

Tested on GKE cluster and found the generated logs in Stackdriver. The
logs stopped when the controllers were idle and once I did actions like
creating namespace or setting parents, the logs started generating
again.

### CODE REVIEW

File: incubator/hnc/pkg/stats/stats.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	HcWrite counter
+
+	// ObRead is the total number of object reads.
+	ObRead counter
+
+	// ObWrite is the total number of object writes.
+	ObWrite counter
+}
+
+var Stats stats
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Not sure if we really need this level of grouping? They could all just be top-level variables, so you'd say `stats.ObjRead.Incr()` instead of `stats.Stats.ObjRead.Incr()` - the second `Stats` doesn't really add anything.

That's assuming you don't want to switch to a more functional style that I recommended above, like `stats.Objects.StartReconcile()` or even just `stats.StartObjectReconcile()`.

File: incubator/hnc/pkg/controllers/hierarchy_controller.go [hunk 2]
```diff
@@ -499,6 +503,7 @@ func (r *HierarchyReconciler) writeNamespace(ctx context.Context, log logr.Logge
 		return false, nil
 	}
 
+	stats.Stats.HcWrite.Incr()
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/pkg/controllers/hierarchy_controller.go]: Sure, it was tracked separately in the previous commit and here I put them together just because the writes both happen in the HC reconciliation.

Separating them sounds more reasonable!

File: incubator/hnc/pkg/controllers/setup.go [hunk 3]
```diff
@@ -12,15 +11,7 @@ import (
...
-var (
-	hcTot int32
-	hcCur int32
-	obTot int32
-	obCur int32
-)
-
-// The ex map is only used by reconcile counters for performance testing. We explicitly
+// The ex map is used by controllers to exclude namespaces to reconcile. We explicitly
 // exclude some default namespaces with constantly changing objects.
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/pkg/controllers/setup.go]: sure

File: incubator/hnc/pkg/stats/stats.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, cur counter = 0, 0
+	working := false
+	actN := 1
+	go func() {
+		// run forever
+		for {
+			// Log activity only when the controllers were still working in the last 0.5s.
+			time.Sleep(500 * time.Millisecond)
+			if Stats.HcTot+Stats.ObTot != total || cur != 0 {
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: I find this logic a bit hard to read. Also, why are we comparing the _current_ values for the totals, but the _last_ values for `cur`?

Maybe rename `total` to `lastTotal`, and then start this block as follows:

```
total = HcTot + ObjTot
cur = HcCur + ObjCur
if total != lastTotal || cur > 0 {
  ...
}
lastTotal = total
```

File: incubator/hnc/pkg/stats/stats.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	ObRead counter
+
+	// ObWrite is the total number of object writes.
+	ObWrite counter
+}
+
+var Stats stats
+
+// LogActivity generates logs for performance testing.
+func LogActivity() {
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: sg

File: incubator/hnc/pkg/stats/stats.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+			// Log activity only when the controllers were still working in the last 0.5s.
+			time.Sleep(500 * time.Millisecond)
+			if Stats.HcTot+Stats.ObTot != total || cur != 0 {
+				// If the controller was previously idle, change its status and log it's started.
+				if working == false {
+					working = true
+					msg := "Activity-" + strconv.Itoa(actN) + "-started"
+					logActivity(log, msg, Stats)
+				}
+				logActivity(log, "Activity", Stats)
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: I'm wondering if "activity" should be one of the structured fields in the log, and then also have a "status" field that can be either "start," "continue" or "finish". That'll make the logs more uniform and easier to grep, I think.

File: incubator/hnc/pkg/controllers/hierarchy_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrianludwin) [reply incubator/hnc/pkg/controllers/hierarchy_controller.go]: I think there will be many cases where we write one and not the other, though this isn't true for reads!

File: incubator/hnc/pkg/controllers/object_controller.go [hunk 7]
```diff
@@ -50,12 +50,14 @@ type ObjectReconciler struct {
...
-		atomic.AddInt32(&obTot, 1)
-		atomic.AddInt32(&obCur, 1)
-		defer atomic.AddInt32(&obCur, -1)
+	if ex[req.Namespace] {
+		return ctrl.Result{}, nil
 	}
 
+	stats.Stats.ObTot.Incr()
+	stats.Stats.ObCur.Incr()
+	defer stats.Stats.ObCur.Decr()
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/pkg/controllers/object_controller.go]: It's definitely more observable to have numbers by object types, but do we just need the logs for performance testing? Since the test case will only have 1 type created in all namespaces.

[the remark this answers is not in the document]
Comment (adrianludwin) [reply incubator/hnc/pkg/controllers/object_controller.go]: Just discussed this offline but for the record - we'll probably use these same calls to generate metrics so it makes sense to group them by GVK too. Actually, on second thought, we don't really need the version, so just group+kind is probably best.

**Comment by adrianludwin** [comment]
As discussed offline: maybe let's be a bit less terse with our variable names, unless they're purely inside the stats package. So for example, `stats.HcWrite.Incr()` is a bit confusing to see in the reconciler; I'd rather see something like `stats.WriteHierarchyConfig()`. Ditto in the logs - `HierarchyConfigWrites` is more self-descriptive than `HcWrites`. But any variable that's purely internal can be short as long as you document them.

File: incubator/hnc/pkg/stats/stats.go (hunk 6 above)
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: In the performance testing scripts, since the logs are grep by intervals (e.g. 5s), the activity start and finish have to be numbered to be able to map to the testing actions. Currently it's grep by e.g. "Activity-$i-finished".

Do you mean generating "activity" number the same way as generating "rid"? It does look more uniform. I will make the changes in this first and figure out how to grep the log in the next PR?

File: incubator/hnc/pkg/controllers/hierarchy_controller.go [hunk 8]
```diff
@@ -529,6 +534,7 @@ func (r *HierarchyReconciler) updateObjects(ctx context.Context, log logr.Logger
 
 // getSingleton returns the singleton if it exists, or creates an empty one if it doesn't.
 func (r *HierarchyReconciler) getSingleton(ctx context.Context, nm string) (*api.HierarchyConfiguration, error) {
+	stats.Stats.HcRead.Incr()
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/pkg/controllers/hierarchy_controller.go]: Sure, just for HC Reconcile, it's supposed to do exact two reads, 1 HC read and 1 NS read.

File: incubator/hnc/pkg/controllers/object_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/pkg/controllers/object_controller.go]: After discussion offline, I will break it by object type for not only the performance testing but also later metrics.

File: incubator/hnc/pkg/stats/stats.go (hunk 6 above)
Comment (adrianludwin) [reply incubator/hnc/pkg/stats/stats.go]: Yeah, pretty much. You could probably do a grep like:

```
cat log | grep "\"activity\": \"5\"" | grep "\"status\": \"finished\"
```

which isn't really as pretty but is definitely more in keeping with how structured logs usually work. wdyt?

File: incubator/hnc/pkg/stats/stats.go (hunk 1 above)
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: I did more abstraction to have `stats.StartObjReconcile(gk)`, `stats.StartHierConfigReconcile()`, `stats.WriteHierConfig()`, etc. in the new change and kept `stats` to have `objects` field with `map[gk]*object` type.

File: incubator/hnc/pkg/stats/stats.go (hunk 4 above)
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: I used the `lastCur != 0` to still output the log for what happened in the past 0.5s. Otherwise, the last active log will be missing.

File: incubator/hnc/pkg/controllers/object_controller.go [hunk 9]
```diff
@@ -50,12 +50,13 @@ type ObjectReconciler struct {
...
 func (r *ObjectReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	if !ex[req.Namespace] {
-		atomic.AddInt32(&obTot, 1)
-		atomic.AddInt32(&obCur, 1)
-		defer atomic.AddInt32(&obCur, -1)
+	if ex[req.Namespace] {
+		return ctrl.Result{}, nil
 	}
 
+	stats.StartObjReconcile(r.GVK.GroupKind())
```
Comment (adrianludwin) [inline incubator/hnc/pkg/controllers/object_controller.go]: Nit: it's sometimes better to make the _caller_ look simpler. So you could say `stats.StartObjReconcile(r.GVK)` and just call `.GroupKind()` inside the function.

There's another school of thought that says "it's better to pass in the minimum amount of information so you don't accidentally come to depend on things (like Version) that you shouldn't care about." In this situation, I'd say that's unlikely to happen so I'd prioritize making the call sites simpler - but it's up to you, this is just a nit.

File: incubator/hnc/pkg/controllers/setup.go [hunk 10]
```diff
@@ -12,19 +11,12 @@ import (
...
 var ex = map[string]bool{
```
Comment (adrianludwin) [inline incubator/hnc/pkg/controllers/setup.go]: Add a TODO comment to make this configurable

File: incubator/hnc/pkg/stats/stats.go [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	i := int32(*c)
+	atomic.AddInt32(&i, 1)
+	*c = counter(i)
+}
+
+func (c *counter) decr() {
+	i := int32(*c)
+	atomic.AddInt32(&i, -1)
+	*c = counter(i)
+}
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Nit: I like to keep the methods of a type close to the definition of the type. I noticed you used to do that and then moved them here? You could also move these to counter.go if you don't want to clutter up the main file.

File: incubator/hnc/pkg/stats/stats.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+func StopHierConfigReconcile() {
+	stats.curHierConfigReconciles.decr()
+}
+
+// StartObjReconcile updates the stats for objects with common GK
+// when an object reconciliation starts.
+func StartObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Why is this needed? Don't we always call `initStats` before starting the controllers?

File: incubator/hnc/pkg/stats/stats.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/go-logr/logr"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/config"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type counter int32
+
+type object struct {
+	gk              schema.GroupKind
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Is this ever used?

File: incubator/hnc/pkg/stats/stats.go [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+// when an object reconciliation finishes.
+func StopObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
+	if _, ok := stats.objects[gk]; !ok {
+		stats.objects[gk] = &object{
+			gk: gk,
+		}
+	}
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: This shouldn't be needed because there's no way to call `StopXXX` without first calling `StartXXX`, right?

File: incubator/hnc/pkg/stats/stats.go [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+	stats.objects[gk].apiWrites.incr()
+}
+
+func initStats() {
+	objects := make(map[schema.GroupKind]*object)
+	for _, gvk := range config.GVKs {
+		objects[gvk.GroupKind()] = &object{
+			gk: gvk.GroupKind(),
+		}
+	}
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Since GVKs will become configurable and changeable in the future, and since you already have code to do this initialization in `StartXXX`, can we remove this?

File: incubator/hnc/pkg/stats/stats.go [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+
+// StartLoggingActivity generates logs for performance testing.
+func StartLoggingActivity() {
+	initStats()
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, lastTotal, lastCur counter = 0, 0, 0
+	working := false
+	go func() {
+		// run forever
+		for {
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Suggestion: factor this out into its own function, then you can get rid of two extra levels of indentation.

File: incubator/hnc/pkg/stats/stats.go [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+			time.Sleep(500 * time.Millisecond)
+			total = stats.totalHierConfigReconciles + getTotalObjReconciles()
+			// If lastCur is not 0 yet, still generate a log for the past 0.5s.
+			if total != lastTotal || lastCur != 0 {
+				// If the controller was previously idle, change its status and log it's started.
+				if working == false {
+					working = true
+					logActivity(log, "start")
+				}
+				logActivity(log, "continue")
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Should this be an `else`?

File: incubator/hnc/pkg/stats/stats.go [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+			} else {
+				// If the controller was previously working, change its status and log it's finished.
+				if working == true {
+					working = false
+					logActivity(log, "finish")
+					stats.actionID.incr()
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: `actionID` doesn't _need_ to be a `counter` because it's only ever accessed from one thread. So you could just say `stats.actionID++` here if you wanted. Up to you.

File: incubator/hnc/pkg/stats/stats.go [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+				}
+			}
+			lastTotal = stats.totalHierConfigReconciles + getTotalObjReconciles()
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Maybe `lastTotal = total`?

File: incubator/hnc/pkg/stats/stats.go [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+			lastCur = stats.curHierConfigReconciles + getCurObjReconciles()
+		}
+	}()
+}
+
+func logActivity(log logr.Logger, status string) {
+	log.Info("Activity", "Action", stats.actionID, "Status", status, "HierConfigWrites", stats.hierConfigWrites, "NamespaceWrites", stats.namespaceWrites, "ObjectWrites", getObjWrites(), "TotalHierConfigReconciles", stats.totalHierConfigReconciles, "CurHierConfigReconciles", stats.curHierConfigReconciles, "TotalObjReconciles", getTotalObjReconciles(), "CurObjReconciles", getCurObjReconciles())
```
Comment (adrianludwin) [inline incubator/hnc/pkg/stats/stats.go]: Nit: maybe format this to make it easier to read?

```
log.Info("Activity",
  "Action", stats.actionID,
  "Status", status,
  ...
)
```

**Comment by rjbez17** [comment]
This lgtm after comments from @adrianludwin  are addressed

File: incubator/hnc/pkg/stats/stats.go (hunk 14 above)
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: yes removed.

File: incubator/hnc/pkg/stats/stats.go (hunk 15 above)
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: agree, removed.

File: incubator/hnc/pkg/stats/stats.go (hunk 12 above)
Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: I got error if I remove this part:
```
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map
```
It's true that `stats.StartLoggingActivity()` is called before creating controllers. Maybe compiler cannot tell?

Comment (adrianludwin) [reply incubator/hnc/pkg/stats/stats.go]: That's a runtime error, not a compiler one. Which goroutine is generating that panic?

Comment (adrianludwin) [reply incubator/hnc/pkg/stats/stats.go]: Ah - if `testLog` is false, `stats.StartLoggingActivity()` is never called - maybe that's what you're seeing? If so, we can probably just say `if stats.objects == nil {return}` instead of creating it.

Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: ah right, exactly! ok will do that.

Comment (yiqigao217) [reply incubator/hnc/pkg/stats/stats.go]: This is done. I added a `testLogEnabled` flag and it's set to true in `stats.StartLoggingActivity()` func.

In each stats update func, it will check the flag first by:
```
if !testLogEnabled {
    return
}
```
